### ISSUE TITLE
Rework enum monomorphization

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#7119]
Currently an enum like `Option` get monomorphized for `Int32` and `Char` like so (`None` is actually called `Option$42.None`)
```
enum Option$42 {
    case None
    case Some(Int32)
}

enum Option$72 {
    case None
    case Some(Char)
}
```

Instead we can abuse welltypedess to conceptually just add specialized cases to a non-parametric enum like so
```
enum Option {
    case None
    case Some$42(Int32)
    case Some$72(Char)
    // and in recursive cases
    case Some$123(Option)
}
```

This achieves
1. No case duplication, e.g. `None` in the above example
2.  Allows monomorphization of infinite specializations (see #6850)
3. Reduces the number of enum types (no duplication in types, only tags)

**Comment by magnus-madsen** [comment]
> @magnus-madsen what do you think of this? I think its a nice representation that doesn't require a lot of refactoring - no new types etc.

Where in the pipeline do you propose to add it?

How many constructors would there be in total?

**Comment by JonathanStarup** [comment]
It would happen in monotypes. there would be less than theres is currently among all the specializations of a given enum. And theyre still generated independently so the amount of cases in one enum shouldnt matter

**Comment by magnus-madsen** [comment]
I think you are proposal is sensible.

But why is it that we cannot simply get rid of enums and the monotypes phase?

We don't have anything like it for say tuples.

**Comment by JonathanStarup** [comment]
hmm i think the only issue is the `Is` expression. even though the term is of the same type, two cases can be different

**Comment by magnus-madsen** [comment]
> hmm i think the only issue is the `Is` expression. even though the term is of the same type, two cases can be different
> 
> but youre right in that a case is just a string attached to an underlying type

So is it a question of giving a unique name to the type? Thats essentially what we need?

**Comment by magnus-madsen** [comment]
For tuples we just compute the name. For tags, in theory, we could just say `Some$Int32`, but of course the names could become long...

**Comment by JonathanStarup** [comment]
Thats why I'm thinking `Some$42`. I agree that the overarching enum itself shouldn't exist, but I think this is a good first step

**Comment by magnus-madsen** [comment]
When we collect types in Reducer or thereabouts, perhaps we could collect all tag types.

Assume we find `Some(Int32)` and `Some(String)`. We put them in a set and then we simply number them. So we get `Some1(Int32)` and `Some2(String)`.

In any case, I am OK with your refactoring as a step towards a better implementation.

But possibly we can go directly for the ball.

**Comment by JonathanStarup** [comment]
Ill rewrite the main body

**Comment by JonathanStarup** [comment]
fixed by #7179
